zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/DedicatedZkClient.java

```diff
@@ -457,17 +470,17 @@ public PathBasedZkSerializer getZkSerializer() {
    * @return
    */
   private void checkIfPathContainsShardingKey(String path) {
-    // TODO: replace with the singleton MetadataStoreRoutingData
     try {
-      String zkRealmForPath = _metadataStoreRoutingData.getMetadataStoreRealm(path);
-      if (!_zkRealmAddress.equals(zkRealmForPath)) {
-        throw new IllegalArgumentException("Given path: " + path + "'s ZK realm: " + zkRealmForPath
-            + " does not match the ZK realm: " + _zkRealmAddress + " and sharding key: "
-            + _zkRealmShardingKey + " for this DedicatedZkClient!");
+      String targetShardingKey = _metadataStoreRoutingData.getShardingKeyInPath(path);
+      if (!_zkRealmShardingKey.equals(targetShardingKey)) {
```

Comment:
Nit, Shall we move this condition check out of try...catch block? I understand it not may be unrelated to your change. But since you already touch it and change the name :)

One more thing: This could throw NPE. `_zkRealmShardingKey` could be null if it is not set in connectionConfig when the user expects it to be a multi-realm mode by default.

Reply:
I am glad that you brought this up. In fact, DedicatedZkClient and SharedZkClient are single-realm APIs, so we need to check in the constructor that Zk Sharding key is not null or empty.

I will add this check. Then we don't need to worry about NPEs.